spyderlib/widgets/sourcecode/base.py

```diff
@@ -811,12 +811,17 @@ def __move_line_or_selection(self, after_current_line=True):
             start_pos = cursor.position()
             cursor.movePosition(QTextCursor.NextBlock)
             end_pos = cursor.position()
+            # check if on last line
+            if end_pos == start_pos:
+                cursor.movePosition(QTextCursor.End)
+                cursor.insertText(os.linesep)

```

Comment:
I can't see a way around adding a new line.  Otherwise, moving `g` up just adds `f` to the same line as `g`.  I fixed point 3 though.